include/hdf5/utils.h

```diff
@@ -0,0 +1,83 @@
+//
+// ICRAR - International Centre for Radio Astronomy Research
+// (c) UWA - The University of Western Australia, 2017
+// Copyright by UWA (in the framework of the ICRAR)
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with this program.  If not, see <https://www.gnu.org/licenses/>.
+//
+
+/**
+ * @file
+ *
+ * Various utilities for interacting with the HDF5 API
+ */
+
+#ifndef SHARK_HDF5_UTILS_H
+#define SHARK_HDF5_UTILS_H
+
+#include <string>
+#include <stdexcept>
+#include <hdf5.h>
+
+namespace shark {
+namespace hdf5 {
+
+void assertHdf5Return(herr_t ret);
+
+template<typename T>
+T assertNonNegative(T val) {
```

Comment:
You'll notice calls to these sprinkled all throughout the codebase, the official API used to handle checking all these return values for us previously. I was trying to avoid a proliferation of if statements inline, but re-reviewing this now as I write this comment (and looking at how the official API handles it) I feel that would perhaps be a more idiomatic way of doing it (and is probably clearer too).

Reply:
I've replaced this with a custom exception + inline checks & throws - much cleaner!

Reply:
(Just a thought, no need to change anything)

In other projects I've used an idiom along the lines of:

```cpp
tempalte <typename Function, typename ... Args>
return_type hdf5_call(Function &&hdf5_function, Args ... args)
{
    if (hdf5_function(std::forward<Args...>(args, ...) < 0) {
        throw exception();
    }
}
```

So then on the calling site you do `hdf5_call(H5Awrite, x, y, z...)` and don't need to check the return value every time. If you want custom error messages that contain the function name you then wrap the `hdf5_call` function in a macro so you can expand the function name as a string constant.

Reply:
Oh that's a nice little idiom, I'll tuck that one away for future reference :smile: But yeah, will leave as is for now - there's a bit of inconsistency in the return types & values (sometime `0` is invalid, others its `<0`, and others have a macro - though they usually resolve to `0`). I think there's some merit to keeping the checks inline to reflect that it's function call dependenct.